include/quic/connection.hpp

```diff
@@ -13,137 +13,157 @@
+
+        - make retransmit timer private
+        - anything touched by network and endpoint is public
+            - make everything private
+            - anything the user calls are public overrides in connection_interface
+        
+    */
+
+    class connection_interface
+    {
+      public:
+        virtual std::shared_ptr<Stream> get_new_stream(
+                stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr) = 0;
+
+        virtual ConnectionID get_conn_id() = 0;
 
-    class Connection
+        // virtual config_t get_user_config() = 0;
+    };
+
+    class Connection : public connection_interface, public std::enable_shared_from_this<Connection>
     {
+      public:
+        // Construct and initialize a new inbound/outbound connection to/from a remote
+        //      ep: owning endpoints
+        //      scid: local ("primary") CID used for this connection (random for outgoing)
+        //		dcid: remote CID used for this connection
+        //      path: network path used to reach remote client
+        //		handle: udp handle dedicated to local address
+        //		creds: relevant tls information per connection
+        //		u_config: user configuration values passed in struct
+        //      dir: enum specifying configuration detailts for client vs. server
+        //		hdr: optional parameter to pass to ngtcp2 for server specific details
+        Connection(
+                Endpoint& ep,
+                const ConnectionID& scid,
+                const ConnectionID& dcid,
+                const Path& path,
+                std::shared_ptr<uv_udp_t> handle,
+                std::shared_ptr<TLSCreds> creds,
+                config_t u_config,
+                Direction dir,
+                ngtcp2_pkt_hd* hdr = nullptr);
+        ~Connection();
+
+        static std::shared_ptr<Connection> make_conn(
+                Endpoint& ep,
+                const ConnectionID& scid,
+                const ConnectionID& dcid,
+                const Path& path,
+                std::shared_ptr<uv_udp_t> handle,
+                std::shared_ptr<TLSCreds> creds,
+                config_t u_config,
+                Direction dir,
+                ngtcp2_pkt_hd* hdr = nullptr);
+
+        // Returns a pointer to the owning endpoint, else nullptr
+        Endpoint* endpoint();
+        const Endpoint* endpoint() const;
+
+        void io_ready();
+
+        const GNUTLSSession* get_session()
+        { return dynamic_cast<GNUTLSSession*>(tls_session.get()); };
+
+        std::shared_ptr<Stream> get_new_stream(
+                stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr) override;
+
+        ConnectionID get_conn_id() override
+        { return _source_cid; }
+
+        Direction direction()
+        { return dir; }
+
+        bool is_closing()
+        { return closing; }
+        bool is_draining()
+        { return draining; }
+        const ConnectionID scid()
```

Comment:
Since ConnectionID is non-trivial, better to return these as `const ConnectionID&` references so that we don't have to copy when doing simple access.  (Likewise for dcid(), path()).

Reply:
and `get_conn_id()`